ps_categorytree.php

```diff
@@ -318,67 +258,107 @@ public function getConfigFieldsValues()
+                $rootCategory = $this->getCurrentCategory();
+                if (!$rootCategory->getSubCategories($rootCategory->id, true)) {
+                    $rootCategory = $this->tryToGetParentCategoryIfAvailable($rootCategory);
                 }
                 break;
-        }
-
-        /* Fallback if something went wrong */
-        if (empty($rootCategory)) {
-            $rootCategory = $homeCategory;
+            default:
+                $rootCategory = $this->getHomeCategory();
         }
 
         return [
             'categories' => $this->getCategories($rootCategory),
             'currentCategory' => $rootCategory->id,
         ];
     }
+
+    /*
+     * Tries to retrieve current category from the context. In case of category controller, it's the category.
+     * In case of product controller, it's either the default category of the product, or the category the customer
+     * came from. This is resolved by the ProductController.
+     */
+    private function getCurrentCategory()
+    {
+        /*
+         * We check several things:
+         * If the controller has the method
+         * If we are on the correct controller
+         * If we have some sensible data and the category is properly loaded
+         */
+        if (
+            !method_exists($this->context->controller, 'getCategory') ||
+            (!$this->context->controller instanceof CategoryController && !$this->context->controller instanceof ProductController) ||
+            empty($this->context->controller->getCategory()) ||
+            !Validate::isLoadedObject($this->context->controller->getCategory())
+        ) {
+            return $this->getHomeCategory();
+        }
+
+        return $this->context->controller->getCategory();
+    }
+
+    /*
+     * Tries to get a parent of the current category.
+     * If we are already on the top of the tree, it will return the input.
+     *
+     * Three cases can happen:
+     * - This category has a normal active parent
+     * - This category has a disabled parent
+     * - This category is already the home category
+     */
+    private function tryToGetParentCategoryIfAvailable($category)
+    {
+        // If we are already on the top of the tree, nothing to do here
+        if ($category->is_root_category || !$category->id_parent || $category->id == Configuration::get('PS_HOME_CATEGORY')) {
+            return $category;
+        }
+
+        // We try to load the parent
+        $parentCategory = new Category($category->id_parent, $this->context->language->id);
+
+        // If it's disabled or the customer does not have access there, we load the next parent
+        if (!$parentCategory->active || !$category->checkAccess((int) $this->context->customer->id)) {
+            return $this->tryToGetParentCategoryIfAvailable($parentCategory);
+        }
+
+        return $parentCategory;
+    }
+
+    private function getIdsOfCategoriesInPathToCurrentCategory()
```

Comment:
```suggestion
    private function getIdsOfCategoriesInPathToCurrentCategory(): array
```

You can use return types, etc.